### PR TITLE
Add Android caching smoke test

### PR DESCRIPTION
**Pull request by lptr**
Using Santa Tracker

### CODE REVIEW

**Comment by lptr** [comment]
Build: https://builds.gradle.org/viewQueued.html?itemId=9500283

**Comment by lptr** [comment]
Build: https://builds.gradle.org/viewQueued.html?itemId=9518391

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/AndroidCachingSmokeTest.groovy [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        clone.directory = projectDir
+        clone.cloneSubmodules = true
+        clone.call()
+
+        // TODO Log branch / commit ID used in test
+
+        expect:
+        runner(
+                "check",
+                "-Dorg.gradle.android.test.gradle-installation=" + System.getProperty('integTest.gradleHomeDir'),
```
Comment (wolfs) [inline subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/AndroidCachingSmokeTest.groovy]: I guess we cannot use `IntegrationTestBuildContext` in smoke tests, right?

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/AndroidCachingSmokeTest.groovy [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                "-Dorg.gradle.android.test.show-output=true"
+            )
+            .withProjectDir(projectDir)
+            .forwardOutput()
+            .build()
+
+        cleanup:
+        projectDir.deleteDir()
```
Comment (wolfs) [inline subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/AndroidCachingSmokeTest.groovy]: Why do you need to do the cleanup? The `TemporaryFolderRule` would delete the project dir anyway, right?

**Comment by oehme** [comment]
It would be nice if instead of hardcoding the expected outcomes we could compare a non-relocated run and a relocated run. I.e. run `assembleDebug` to fill the cache, run `clean assembleDebug` to create the expectations, then run `assembleDebug` in the other directory to check the expectations.

**Comment by lptr** [comment]
@oehme The problem with deriving the expectations automatically from the non-relocated case is that we might do something in Gradle (or Google might do something in Android) that breaks caching for the non-relocated case as well. If this happened, the automated expectations wouldn't catch it.

What we could do is to also run the test in non-relocated mode, and perhaps also without `clean`. That way we could make it explicit which task is expected to work with incremental build, which with non-relocatable caching, and which with relocatable. There should be no difference between the latter two, but unfortunately there is currently.

Another dimension I'd like to add is what happens without the cache-fix plugin. The ultimate goal is for all cached tasks with and without the fix plugin should be cacheable.

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/AndroidCachingSmokeTest.groovy (hunk 1 above)
Comment (lptr) [reply subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/AndroidCachingSmokeTest.groovy]: The super-class also uses this approach. I could extract it into a method, though.

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/AndroidCachingSmokeTest.groovy (hunk 2 above)
Comment (lptr) [reply subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/AndroidCachingSmokeTest.groovy]: Good point, will remove this.

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/AndroidCachingSmokeTest.groovy (hunk 1 above)
Comment (lptr) [reply subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/AndroidCachingSmokeTest.groovy]: It is actually available, fixed.

**Comment by lptr** [comment]
The TeamCity build is fixed, and I addressed your comments @wolfs. Please take another look.

**Comment by lptr** [comment]
Another build, now testing both with and without cache fix plugin: https://builds.gradle.org/viewQueued.html?itemId=9564515

**Pull request merged**